fiona/io.py

```diff
@@ -52,33 +52,33 @@ def open(self, driver=None, schema=None, crs=None, encoding=None,
         if self.closed:
             raise IOError("I/O operation on closed file.")
 
-        if not self.exists():
-            self._ensure_extension(driver)
-            this_schema = schema.copy()
-            this_schema["properties"] = OrderedDict(schema["properties"])
-            return Collection(
-                self.name,
-                "w",
-                crs=crs,
-                driver=driver,
-                schema=this_schema,
-                encoding=encoding,
-                layer=layer,
-                enabled_drivers=enabled_drivers,
-                crs_wkt=crs_wkt,
-                **kwargs
-            )
-
-        elif self.mode in ("r", "r+"):
-            return Collection(
-                self.name,
-                "r",
-                driver=driver,
-                encoding=encoding,
-                layer=layer,
-                enabled_drivers=enabled_drivers,
-                **kwargs
-            )
+        if driver is not None and not supports_vsi(driver):
+            raise DriverError("Driver {} does not support virtual files.")
+
+        if mode in ('r', 'a') and not self.exists():
+            raise IOError("MemoryFile does not exist.")
+        if layer is None and mode == 'w' and self.exists():
+            raise IOError("MemoryFile already exists.")
+
+        if not self.exists() or mode == 'w':
+            if driver is not None:
```

Comment:
Should we introduce a check that driver and schema need to be specified in write mode?